compiler/rustc_mir_transform/src/deduce_param_attrs.rs

```diff
@@ -0,0 +1,93 @@
+//! Deduces supplementary parameter attributes from MIR.
+//!
+//! Deduced parameter attributes are those that can only be soundly determined by examining the
+//! body of the function instead of just the signature. These can be useful for optimization
+//! purposes on a best-effort basis. We compute them here and store them into the crate metadata so
+//! dependent crates can use them.
+
+use rustc_hir::def_id::DefId;
+use rustc_index::bit_set::BitSet;
+use rustc_middle::mir::visit::{PlaceContext, Visitor};
+use rustc_middle::mir::{Body, Local, Location};
+use rustc_middle::ty::{DeducedParamAttrs, ParamEnv, TyCtxt};
+use rustc_span::DUMMY_SP;
+
+// A visitor that determines which arguments have been mutated. We can't use the mutability field on
+// LocalDecl for this because it has no meaning post-optimization.
```

Comment:
Nit : make all of these doc comments as we document private items in nightly docs